interfaces/builtin/systemd-control.go

```diff
@@ -0,0 +1,62 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+)
+
+const systemdControlConnectedPlugAppArmor = `
+# Description: Can control systemd
+
+/bin/systemctl ix,
+
+#include <abstractions/dbus-strict>
+
+dbus (send,receive)
+    bus=system
+    path=/org/freedesktop/systemd1{,/**}
+    interface=org.freedesktop.DBus.*
+    peer=(label=unconfined),
+dbus (send,receive)
+    bus=system
+    path=/org/freedesktop/systemd1{,/**}
+    interface=org.freedesktop.systemd1.*
+    peer=(label=unconfined),
+`
```

Comment:
Does this policy actually work? I tried the policy as in this PR and needed to add:
```
/run/systemd/private rw,
ptrace (trace) peer=unconfined,
/{,usr/}bin/systemd-tty-ask-password-agent ixr,
```

and get:
```
bash-4.3# systemctl stop snapweb
Running in chroot, ignoring request.
```

I then tried to see if I could refine the policy to only what was needed using this:
```
/{,usr}/sbin/service ixr,
/etc/systemd/system/{,**} r,
/run/systemd/system/{,**} r,

/{,usr/}bin/systemctl ixr,
/{,usr/}bin/systemd-tty-ask-password-agent ixr,

#include <abstractions/dbus-strict>

# Allow listing all units
dbus (send)
    bus=system
    path=/org/freedesktop/systemd1
    interface="org.freedesktop.systemd1.Manager"
    member=ListUnits{,Filtered}
    peer=(label=unconfined),

dbus (send)
    bus=system
    path=/org/freedesktop/systemd1
    interface="org.freedesktop.systemd1.Manager"
    member={Start,Stop,Restart}Unit
    peer=(label=unconfined),

dbus (send)
    bus=system
    path="/org/freedesktop/systemd1/unit/*"
    interface="org.freedesktop.DBus.Properties"
    member=Get
    peer=(label=unconfined),

# This rule is unacceptable
ptrace (trace) peer=unconfined,

capability net_admin,
# TODO: what does this allow?
/run/systemd/private rw,
```

and got the same result. This is on classic Ubuntu. Part of this I think is that polkit is getting in the way as such, this interface as written doesn't work quite right.

The required ptrace rule also grants more privilege than is expected by this interface and allows this snap to ptrace and control all unconfined processes.